From 2093a9abe63806224d695fad817dc4759c1a7078 Mon Sep 17 00:00:00 2001
From: Cliff Chen <cliff.chen@rock-chips.com>
Date: Mon, 16 Jul 2018 14:40:48 +0800
Subject: [PATCH] f2fs: add a new limit for reserve root

The reserved root blocks is not enough for booting Android due to
the limit of 0.2% if the fs size too small. so we add a new mini-
mum limit is 128MB.

Change-Id: I5af3b182001d27e4d18b4090c5270bbb2ac6253b
Signed-off-by: Cliff Chen <cliff.chen@rock-chips.com>
---
 fs/f2fs/f2fs.h  | 2 ++
 fs/f2fs/super.c | 3 ++-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/fs/f2fs/f2fs.h b/fs/f2fs/f2fs.h
index 97c17b3d984c..4587fbadc555 100644
--- a/fs/f2fs/f2fs.h
+++ b/fs/f2fs/f2fs.h
@@ -70,6 +70,8 @@ extern char *fault_name[FAULT_MAX];
 #define IS_FAULT_SET(fi, type) ((fi)->inject_type & (1 << (type)))
 #endif
 
+#define MIN_ROOT_RESERVED_BLOCKS (128 * 1024 * 1024)
+
 /*
  * For mount options
  */
diff --git a/fs/f2fs/super.c b/fs/f2fs/super.c
index cb57ad3ca32d..dcce72aa47a6 100644
--- a/fs/f2fs/super.c
+++ b/fs/f2fs/super.c
@@ -211,7 +211,8 @@ static inline void limit_reserve_root(struct f2fs_sb_info *sbi)
 
 	/* limit is 0.2% */
 	if (test_opt(sbi, RESERVE_ROOT) &&
-			F2FS_OPTION(sbi).root_reserved_blocks > limit) {
+			F2FS_OPTION(sbi).root_reserved_blocks > limit &&
+			F2FS_OPTION(sbi).root_reserved_blocks > MIN_ROOT_RESERVED_BLOCKS) {
 		F2FS_OPTION(sbi).root_reserved_blocks = limit;
 		f2fs_msg(sbi->sb, KERN_INFO,
 			"Reduce reserved blocks for root = %u",
-- 
2.35.3

